### ISSUE TITLE
Update SsnValidator implement

### ISSUE THREAD

**Issue by Soullivaneuh** [Soullivaneuh/IsoCodesValidator#6]
If https://github.com/ronanguilloux/IsoCodes/pull/34 is merged, review `SsnValidator` class to handle singleton class implementation.

**Comment by enumag** [comment]
Singleton is wrong, you should use Ssn as service in DIC and require it in constructor of SsnValidator.

**Comment by Soullivaneuh** [comment]
That's an idea, but I don't want to require symfony/dependency-injection here because a provide both Symfony and Silex support.

**Comment by enumag** [comment]
You should NOT require symfony/dependency-injection. The SsnValidator doen't care where the instance will come from. And you can use a fallback:

```
public function __construct(IsoCodes\Ssn $ssn = null)
{
    $this->ssn = $ssn ?: new IsoCodes\Ssn();
}
```

This is the correct solution in my opinion. The singleton PR should not get merged.

**Comment by Soullivaneuh** [comment]
I don't get it. From where the `Ssn` instance passed on the constructor come from then?

And BTW, singleton system also have goal to match the `IsoCodesInterface` later: https://github.com/ronanguilloux/IsoCodes/issues/33

**Comment by enumag** [comment]
I'm not sure how symfony/dependency-injection works but presumably when you want to add a service you write some kind of extension class right? You should add such class to this package for those users that use symfony/dependency-injection. But you should not make it a dependency as it is not an essential part of this package. I could implement an extension for nette/di if you want me to.

Oh and singletons are plain wrong pretty much everywhere.

**Comment by enumag** [comment]
Oh and since you still don't understand the optional dependencies, [this package](https://github.com/cocur/slugify) is a nice example of what I have in mind. If you look into composer.json it doesn't have any dependencies. Yet it [contains](https://github.com/cocur/slugify/tree/master/src/Bridge) classes which use parts of Symfony/Twig/Zend/Nette/Latte and god knows what else. None od them is required because these classes are just **bridges**, not essential parts of the package. But you can find them all in require-dev section in composer json because they are obviously nedded for tests.

In theory you should have a separate composer package for every single one of these bridges. It would be difficult to maintain though which is why such solution is not used.

You should do the same here - add a bridge for symfony/dependency-injection and maybe some other packages (or just wait for PRs) but don't add it to dependencies.

**Comment by Soullivaneuh** [comment]
> I could implement an extension for nette/di if you want me to.

Not sure if it will be accepted, but I'm curious to see your POC.

> Oh and singletons are plain wrong pretty much everywhere.

May I ask why? :-)

Honestly, I'm not a big fan of singleton too. I chose this because it's the only way I found to match `IsoCodesInterface` as I said before.

> In theory you should have a separate composer package for every single one of these bridges.

Yeah, this what I did before, but I deprecate them to get all on the same because of pain to maintain 5 lines of codes for each new release:

* https://github.com/Soullivaneuh/SLLHIsoCodesValidatorBundle
* https://github.com/Soullivaneuh/IsoCodesValidatorServiceProvider

**Comment by enumag** [comment]
>> I could implement an extension for nette/di if you want me to.

> Not sure if it will be accepted, but I'm curious to see your POC.

I'd have to study your bundle in more detail to see if there is anything else I should register as service but for the SsnValidator (if you add the constructor I from above) the following extension class would be enough for nette bridge. The Ssn instance from DI container would be passed to SsnValidator because of [this line](https://github.com/Kdyby/Validator/blob/master/src/Kdyby/Validator/ConstraintValidatorFactory.php#L54) in Kdyby/Validator which is an integration for symfony/validator to nette/di. Or I could just add SsnValidator as another service, which is a bit cleaner but doesn't matter.

```php
<?php

namespace SLLH\IsoCodesValidator\Bridge\Nette;

use Nette\DI\CompilerExtension;

class IsoCodesExtension extends CompilerExtension
{

    public function loadConfiguration()
    {
        $builder = $this->getContainerBuilder();

        $builder->addDefinition($this->prefix('ssn'))
            ->setClass('IsoCodes\Ssn');
    }

}
```

>> Oh and singletons are plain wrong pretty much everywhere.

> May I ask why? :-)

My main reason is that when you use a DI container it pretty much solves the need of only having one instance of a class with the benefits that 1) The class itself doesn't know anything about it. 2) In case you really need two instances with different configuration it is achievable.

----

So what do you think about the solution with `SsnValidator::__construct` from above + having bridges for Symfony and perhaps Silex in this package (without explicitly requiring any of them)?

**Comment by Soullivaneuh** [comment]
> In case you really need two instances with different configuration it is achievable.

Well, for this case we really don't need two instances of `Ssn`.

> So what do you think about the solution with SsnValidator::__construct from above + having bridges for Symfony and perhaps Silex in this package (without explicitly requiring any of them)

I also have bridge for Symfony and Silex. See `Bundle` and `Provider` folders. But yes, I think this should be move to `Bridge\[Symfony|Silex]` folder.

For `SsnValidator::__construct` Nette framework, as I said, feel free to propose a PR. :+1: 

> You would not need the singleton that way, right?

If it's integrated, no indeed. But this will still be needed on `IsoCodes` library to match `IsoCodesInterface`.

**Comment by enumag** [comment]
I don't see why singleton would be necessary for `IsoCodesInterface`. Can you elaborate the reason? What should the `IsoCodesInterface` look like?

**Comment by Soullivaneuh** [comment]
@enumag This one: https://github.com/ronanguilloux/IsoCodes/blob/master/src/IsoCodes/IsoCodeInterface.php

If you have a better implementation, feel free to propose a PR on the related repository (I'm not the maintainer of IsoCodes).

**Comment by enumag** [comment]
@Soullivaneuh Well that's easy. Unless I'm missing something which is always possible of course. ;-)

https://github.com/ronanguilloux/IsoCodes/pull/73

**Comment by Soullivaneuh** [comment]
Resolved by #63.
